iroha/src/kura.rs

```diff
@@ -180,126 +196,248 @@ impl Default for Mode {
+                .await
+                .wrap_err("Failed to fetch storage name for block")?;
+        Ok(self.path.join(filename))
     }
 
     /// Write block as file onto the disk.
     #[iroha_futures::telemetry_future]
     pub async fn write(&self, block: &VersionedCommittedBlock) -> Result<Hash> {
-        //filename is its height
-        let path = self.get_block_path(block.header().height);
-        let mut file = File::create(path)
+        let path = self
+            .get_block_path(
+                NonZeroU64::new(block.header().height)
+                    .wrap_err("An attempt to write a zero-height block.")?,
+            )
+            .await?;
+        let mut file = fs::OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(path)
             .await
             .wrap_err("Failed to open storage file.")?;
         let hash = block.hash();
         let serialized_block: Vec<u8> = block.encode_versioned()?;
+        #[allow(clippy::cast_possible_truncation)]
+        let block_size = serialized_block.len() as u32;
+        file.write_u32_le(block_size).await.wrap_err("Failed to write to storage file.")?;
         file.write_all(&serialized_block)
             .await
             .wrap_err("Failed to write to storage file.")?;
         Ok(hash)
     }
 
-    /// Returns a block at height, read from disk.
-    ///
-    /// # Errors
-    /// - There is no file corresponding to this block.
+    /// Returns a sorted vector of blocks starting from 1 height to the top block.
     #[iroha_futures::telemetry_future]
-    pub async fn read(&self, height: u64) -> Result<VersionedCommittedBlock> {
-        let path = self.get_block_path(height);
-        let mut file = File::open(&path).await.wrap_err("No file found.")?;
-        let metadata = metadata(&path).await.wrap_err("Unable to read metadata.")?;
-        #[allow(clippy::cast_possible_truncation)]
-        let mut buffer = vec![0; metadata.len() as usize];
-        let _ = file.read(&mut buffer).await.wrap_err("Buffer overflow.")?;
-        VersionedCommittedBlock::decode_versioned(&buffer)
-            .wrap_err("Failed to read block from store.")
+    pub async fn read_all(&self) -> Result<Vec<VersionedCommittedBlock>> {
+        let mut blocks: Vec<VersionedCommittedBlock> = Vec::new();
+        let base_indices = storage_files_base_indices(&self.path).await?;
+        for base_index in base_indices {
+            let path = self.path.join(base_index.to_string());
+            let mut read_blocks = read_all_from_storage_file(&path)
+                .await
+                .wrap_err("Failed to deserialize blocks.")?;
+            let prev_block = read_blocks
+                .first()
+                .wrap_err(format!(
+                    "No blocks were found in file {}. Data storage might be corrupted.",
+                    path.to_string_lossy()
+                ))?
+                .header()
```

Comment:
In all of the cases where you return errors - you return a boxed string with message. It is ok for high level functions that just forward error up the call stack. But BlockStore is a more low level (library structure) I would say, and we might want to handle some of this errors in the future.

So I would suggest having `BlockStoreError` enum with all variants that you have here. You might use `thiserror` for this. And then for this function for example you would return `Result<Vec<VersionedCommittedBlock>, BlockStoreError>`.

Though this is not critical now I guess if you don't handle this errors anywhere on top. But might be good for the future.

Reply:
Also maybe a good idea will be at least to do `wrap_err_with(|| ...)`